internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+package jobobject
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/processorinfo"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// JobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job, a channel to receive iocp notifications about the lifecycle
+// of the job and a mutex for synchronized handle access.
+type JobObject struct {
+	jobHandle  windows.Handle
+	notifChan  chan interface{}
+	handleLock sync.RWMutex
+}
+
+type jobLimits struct {
+	cpuCount        uint32
+	cpuRate         uint32
+	cpuWeight       uint32
+	memoryLimitInMB uintptr
+	maxIops         int64
+	maxBandwidth    int64
+}
+
+// Processor resource controls
+const (
+	processorLimitMin  = 1
+	processorLimitMax  = 10000
+	processorWeightMin = 0
+	processorWeightMax = 10000
+)
+
+var errAlreadyClosed = errors.New("the handle has already been closed")
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(ctx context.Context, limits *jobLimits) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.memoryLimitInMB != 0 {
+		var (
+			basicLimitFlags uint32
+			eliInfo         windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION
+		)
+		basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+		if (limits.memoryLimitInMB << 20) > math.MaxUint64 {
+			return fmt.Errorf("memory limit value of `%d` is invalid", limits.memoryLimitInMB)
+		}
+		eliInfo.JobMemoryLimit = (limits.memoryLimitInMB << 20)
+		eliInfo.BasicLimitInformation.LimitFlags = basicLimitFlags
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 || limits.cpuWeight != 0 || limits.cpuCount != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.cpuCount != 0 {
```

Comment:
I would say let's get rid of this `cpuCount` case. The caller can already do the same thing themselves by setting `cpuRate`. I think providing a CPU count control where we just convert it into a rate is confusing.

Reply:
Mostly just following what HCS does for cpu count/rate/weight here as I'd assume we'd want to support the same

Reply:
I don't think there's any reason we need to expose the same interface as what HCS does. I'd like this package to not be opinionated towards the containers use case. Mostly I find the HCS conversions to be weird abstractions that don't really add any value. If a caller does want to do something similar to what HCS does, they can just implement that logic themselves before passing a value to us here, which provides better API layering.